servicetalk-http-api/src/main/java/io/servicetalk/http/api/HttpServerBuilder.java

```diff
@@ -412,20 +445,74 @@ public final ServerContext listenBlockingStreamingAndAwait(
     private Single<ServerContext> listenForService(StreamingHttpService rawService, HttpExecutionStrategy strategy) {
         ConnectionAcceptor connectionAcceptor = connectionAcceptorFactory == null ? null :
                 connectionAcceptorFactory.create(ACCEPT_ALL);
-        StreamingHttpServiceFilterFactory currServiceFilter = serviceFilter;
-        if (!AsyncContext.isDisabled()) {
-            StreamingHttpServiceFilterFactory asyncContextFilter = new AsyncContextAwareHttpServiceFilter();
-            currServiceFilter = currServiceFilter == null ? asyncContextFilter :
-                    appendFilter(asyncContextFilter, currServiceFilter);
+
+        Deque<StreamingHttpServiceFilterFactory> nonOffloadingFilters = new ArrayDeque<>(noOffloadServiceFilters);
+
+        StreamingHttpService filteredService;
+
+        if (nonOffloadingFilters.isEmpty()) {
+            filteredService = serviceFilters.isEmpty() ? rawService : buildService(serviceFilters, rawService);
+        } else {
+            boolean anyOffloads = strategy.isSendOffloaded() ||
+                    strategy.isMetadataReceiveOffloaded() ||
+                    strategy.isDataReceiveOffloaded();
+
+            if (anyOffloads) {
+                // We are going to have to offload, even if just to the raw service
+                nonOffloadingFilters.add(new OffloadingFilter(strategy,
+                        buildFactory(serviceFilters)));
+                strategy = null != strategy.executor() ?
+                        HttpExecutionStrategies.customStrategyBuilder()
+                                .offloadNone().executor(strategy.executor()).build() :
+                        HttpExecutionStrategies.noOffloadsStrategy();
+            } else {
+                // All the filters can be appended.
+                nonOffloadingFilters.addAll(serviceFilters);
```

Comment:
My stream-fu isn't great, so asking: will `Stream.concat(nonOffloadingFilters.stream, serviceFilters.stream())` do better than `addAll` here?

Reply:
Yes, now converted to Stream.